password.js

```diff
@@ -0,0 +1,152 @@
+const crypto = require('crypto');
+const fs = require('fs');
+
+var protobuf = require('protocol-buffers');
+
+/**
+ * The password module provides utilities for dealing with passwords. This
+ * includes hashing and verifying passwords as well as derive secure
+ * cryptographic keys from passwords.
+ * @module easy-crypto/password
+ */
+
+/**
+ * An error that hints that the hashing algorithm used is no longer valid and a
+ * rehash is required.
+ * @constant
+ * @type {Error}
+*/
+const InvalidHashError = new Error('Invalid algorithm, rehash required.');
+
+/**
+ * Derive a cryptographically secure key using a password and a salt.
+ * @param {Data} password The password to be used
+ * for key derivation.
+ * @param {Data} salt The salt to be applied to the
+ * password. The salt should be as unique as possible. It is recommended that a
+ * salt is random and at least 16 bytes long. See NIST SP 800-132 for details.
+ * @param {number} iterations The number of iterations to be performed. The
+ * value must be a number set as high as possible. The higher the number of
+ * iterations, the more secure the derived key will be, but will take a longer
+ * amount of time to complete.
+ * @param {number} keylen The length of the key to be produced.
+ * @param {string} digest The HMAC digest algorithm to be used.
+ * @returns {Promise<Buffer>} The derived key.
+ */
+function deriveKey(password, salt, iterations, keylen, digest) {
+  return new Promise((resolve, reject) => {
+    crypto.pbkdf2(password, salt, iterations, keylen, digest, (err, key) => {
+      if (err) reject(err);
+      resolve(key);
+    });
+  });
+}
+
+/**
+ * Derive a cryptographically secure key synchronously using a password and a salt.
+ * @param {Data} password The password to be used
+ * for key derivation.
+ * @param {Data} salt The salt to be applied to the
+ * password. The salt should be as unique as possible. It is recommended that a
+ * salt is random and at least 16 bytes long. See NIST SP 800-132 for details.
+ * @param {number} iterations The number of iterations to be performed. The
+ * value must be a number set as high as possible. The higher the number of
+ * iterations, the more secure the derived key will be, but will take a longer
+ * amount of time to complete.
+ * @param {number} keylen The length of the key to be produced.
+ * @param {string} digest The HMAC digest algorithm to be used.
+ * @returns {Buffer} The derived key.
+ */
+function deriveKeySync(password, salt, iterations, keylen, digest) {
+  return crypto.pbkdf2Sync(password, salt, iterations, keylen, digest);
+}
+
+/**
+ * Hash a password for storage.
+ * @param {Data} password The password to be hashed.
+ * @returns {Promise<Buffer>} The hashed password optimized for storage.
+ */
+function hashPassword(password) {
+  const salt = crypto.randomBytes(32);
+  return new Promise((resolve, reject) => {
+    fs.readFile('password.proto', (err, data) => {
```

Comment:
Got it. I just wanted to make sure I didn't take any more time than needed for a user that didn't even need to read that file. Also, since mafintosh suggested that the throughput was surprisingly good, I didn't think it'll be too bad. But agreed, I will move it.